lib/fluent/plugin/buffer.rb

```diff
@@ -0,0 +1,351 @@
+      config_param :total_bytes_limit, :size, default: DEFAULT_TOTAL_BYTES_LIMIT
+
+      # If user specify this value and (chunk_size * queue_length) is smaller than total_size,
+      # then total_size is automatically configured to that value
+      config_param :queue_length_limit, :integer, default: nil
+
+      # optional new limitations
+      config_param :chunk_records_limit, :integer, default: nil
+
+      attr_accessor :log
+
+      Metadata = Struct.new(:timekey, :tag, :variables)
+
+      def initialize
+        super
+
+        @chunk_bytes_limit = nil
+        @total_bytes_limit = nil
+        @queue_length_limit = nil
+        @chunk_records_limit = nil
+
+        @log = nil
+
+        @_owner = nil
+        @_plugin_id = nil
+        @_plugin_id_configured = false
+      end
+
+      def persistent?
+        false
+      end
+
+      def configure(conf)
+        super
+
+        if @queue_length_limit && @total_bytes_limit > @chunk_bytes_limit * @queue_length_limit
+          @total_bytes_limit = @chunk_bytes_limit * @queue_length_limit
+        end
+      end
+
+      def start
+        super
+        # stage #=> Hash (metadata -> chunk): not flushed yet
+        # queue #=> Array (chunks)          : already flushed (not written)
+        @stage, @queue = resume
+
+        @queued_num = {} # metadata => int (number of queued chunks)
+        @queue.each do |chunk|
+          @queued_num[chunk.metadata] ||= 0
+          @queued_num[chunk.metadata] += 1
+        end
+
+        @dequeued = {} # unique_id => chunk
+
+        @stage_size = @queue_size = 0
+        @metadata_list = [] # keys of @stage
+      end
+
+      def close
+        super
+        synchronize do
+          @dequeued.each_pair do |chunk_id, chunk|
+            chunk.close
+          end
+          until @queue.empty?
+            @queue.shift.close
+          end
+          @stage.each_pair do |metadata, chunk|
+            chunk.close
+          end
+        end
+      end
+
+      def terminate
+        super
+        @stage = @queue = nil
+      end
+
+      def storable?
+        @total_size_limit > @stage_size + @queue_size
+      end
+
+      ## TODO: for back pressure feature
+      # def used?(ratio)
+      #   @total_size_limit * ratio > @stage_size + @queue_size
+      # end
+
+      def resume
+        # return {}, []
+        raise NotImplementedError, "Implement this method in child class"
+      end
+
+      def generate_chunk(metadata)
+        raise NotImplementedError, "Implement this method in child class"
+      end
+
+      def metadata_list
+        synchronize do
+          @metadata_list.dup
+        end
+      end
+
+      def new_metadata(timekey: nil, tag: nil, variables: nil)
+        Metadata.new(timekey, tag, variables)
+      end
+
+      # TODO: metadata_list cleaner
+      def add_metadata(metadata)
+        synchronize do
+          if i = @metadata_list.index(metadata)
+            @metadata_list[i]
```

Comment:
I confirmed that following code is wrong. Return value is needed to use exact a object for same metadata.